### PR TITLE
osutil: add support for symlinks to EnsureFileState

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This is part of the preparation for using `Ensure*` functions instead of `Add*` functions in `backend.LinkSnap` and `backend.UnlinkSnap`.

This is needed when replacing snap binaries as they are symlinks to `/usr/bin/snap`.
For more context: #13066

**Note: by adding support for symlinks to `EnsureFileSate` the other `Ensure*` functions also grow the feature.**

### CODE REVIEW

File: osutil/syncdir.go [hunk 1]
```diff
@@ -29,9 +29,29 @@ import (
...
+
+const (
+	RegularFileStateType FileStateType = "regular"
+	SymlinkFileStateType FileStateType = "symlink"
+)
+
 // FileState is an interface for conveying the desired state of a some file.
 type FileState interface {
 	State() (reader io.ReadCloser, size int64, mode os.FileMode, err error)
+	Type() FileStateType
```
Comment (pedronis) [inline osutil/syncdir.go:42]: os.FileMode can also carry a type, see ModeSymlink, so not sure if we strictly needs this?

File: osutil/syncdir.go [hunk 2]
```diff
@@ -108,6 +140,7 @@ var ErrSameState = fmt.Errorf("file state has not changed")
 //
 // In all cases, the function returns the first error it has encountered.
 func EnsureDirStateGlobs(dir string, globs []string, content map[string]FileState) (changed, removed []string, err error) {
+	// TODO: add tests for symlinks
```
Comment (ZeyadYasser) [inline osutil/syncdir.go]: remove TODO comment

File: osutil/syncdir.go (hunk 1 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go]: Updated! Thanks for pointing this out.

Comment (ZeyadYasser) [reply osutil/syncdir.go]: I updated to returning `mode.Perms()` in State functions to ignore modes passed by mistake and only consider permissions.

File: osutil/syncdir.go [hunk 3]
```diff
@@ -34,6 +34,14 @@ type FileState interface {
 	State() (reader io.ReadCloser, size int64, mode os.FileMode, err error)
 }
 
+type SymlinkFileState struct {
```
Comment (pedronis) [inline osutil/syncdir.go:37]: this should have a doc comment

File: osutil/syncdir_test.go [hunk 4]
```diff
@@ -241,3 +241,69 @@ func (s *EnsureDirStateSuite) TestRemovesAllManagedFilesOnError(c *C) {
...
+	c.Assert(removed, HasLen, 0)
+	// The symlinks points to original
+	c.Assert(missingSymlink, testutil.FileEquals, "data")
+	c.Assert(osutil.IsSymlink(missingSymlink), Equals, true)
+	link, err := os.Readlink(missingSymlink)
+	c.Assert(err, IsNil)
+	c.Assert(link, Equals, original)
+}
+
+func (s *EnsureDirStateSuite) TestReplaceFileWithSymlink(c *C) {
```
Comment (pedronis) [inline osutil/syncdir_test.go:287]: don't we also need a replace symlink with symlink test? and also a do nothing test?

parts of symlinkFileStateEqualTo are not exercised atm

**Review by pedronis — COMMENTED** [review]
thanks, did another pass. The summary/description should probably explain that by supporting this in EnsureFileState, the other Ensure* functions also grow the feature.

File: osutil/syncdir.go (hunk 3 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go:38]: Done

File: osutil/syncdir_test.go (hunk 4 above)
Comment (ZeyadYasser) [reply osutil/syncdir_test.go:289]: Yes indeed, I missed that, thank you! By a nothing test, I assumed you mean trying to replace a symlink with an equivalent symlink, right?

File: osutil/syncdir.go [hunk 5]
```diff
@@ -34,6 +34,7 @@ type FileState interface {
 	State() (reader io.ReadCloser, size int64, mode os.FileMode, err error)
 }
 
+// SymlinkFileState describes the desired symlink by providing a target
```
Comment (pedronis) [inline osutil/syncdir.go:37]: missing final .   also "a target"/"its target"

**Review by pedronis — APPROVED** [review]
thanks, small doc comment remarks

File: osutil/syncdir.go (hunk 5 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go]: done, thanks!

File: osutil/syncdir_test.go [hunk 6]
```diff
@@ -241,3 +243,134 @@ func (s *EnsureDirStateSuite) TestRemovesAllManagedFilesOnError(c *C) {
...
+		os.ModeDir,
+		os.ModeNamedPipe,
+		os.ModeSocket,
+		os.ModeDevice,
+		os.ModeCharDevice,
+		os.ModeIrregular,
+	}
+	filePath := filepath.Join(s.dir, "test.snap")
+	for _, modeType := range unsupportedModeTypes {
+		fileState := &mockFileState{mode: modeType, err: nil}
```
Comment (MiguelPires) [inline osutil/syncdir_test.go:371]: tiny nitpick: the `err: nil` isn't necessary

File: osutil/syncdir.go [hunk 7]
```diff
@@ -49,7 +58,7 @@ func (fref FileReference) State() (io.ReadCloser, int64, os.FileMode, error) {
 	if err != nil {
 		return nil, 0, os.FileMode(0), err
 	}
-	return file, fi.Size(), fi.Mode(), nil
+	return file, fi.Size(), fi.Mode().Perm(), nil
```
Comment (MiguelPires) [inline osutil/syncdir.go:61]: Why these changes?

**Review by MiguelPires — COMMENTED** [review]
Some questions and comments

File: osutil/syncdir.go (hunk 7 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go:61]: This is needed to ensure that if ModeType was mistaknely passed to either `FileReferencePlusMode` and `MemoryFileState`, it wouldn't have effect (it already doesn't as AtomicWrite ignores mode type anyway).
This is needed to avoid `FileReferencePlusMode` and `MemoryFileState` passed with `os.ModeSymlink` messing this check [here](https://github.com/snapcore/snapd/pull/13071/files#diff-c96c5c4903c95e4dc4fa6b1e968c98ceb20c526551ec6509f6670cdc80489201R314).

There was a discussion about this with @pedronis https://github.com/snapcore/snapd/pull/13071#discussion_r1287066580.

File: osutil/syncdir.go [hunk 8]
```diff
@@ -240,3 +247,72 @@ func EnsureFileState(filePath string, state FileState) error {
...
+	_, _, mode, err := state.State()
+	if err != nil {
+		return err
+	}
+	switch mode.Type() {
+	case 0:
+		return ensureRegularFileState(filePath, state)
+	case os.ModeSymlink:
+		return ensureSymlinkFileState(filePath, state)
+	}
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply osutil/syncdir.go]: updated, thanks!

File: osutil/syncdir_test.go [hunk 9]
```diff
@@ -241,3 +243,134 @@ func (s *EnsureDirStateSuite) TestRemovesAllManagedFilesOnError(c *C) {
...
+		"missing-symlink.snap": &osutil.SymlinkFileState{Target: original},
+	})
+	c.Assert(err, IsNil)
+	// Created file is reported
+	c.Assert(changed, DeepEquals, []string{"missing-symlink.snap"})
+	c.Assert(removed, HasLen, 0)
+	// The symlink points to original
+	c.Assert(missingSymlink, testutil.FileEquals, "data")
+	c.Assert(osutil.IsSymlink(missingSymlink), Equals, true)
+	link, err := os.Readlink(missingSymlink)
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply osutil/syncdir_test.go]: updated

File: osutil/syncdir.go (hunk 7 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go]: done

File: osutil/syncdir.go [hunk 10]
```diff
@@ -84,7 +90,17 @@ type MemoryFileState struct {
...
 // State returns a reader of the in-memory contents of a file, along with other meta-data.
 func (blob *MemoryFileState) State() (io.ReadCloser, int64, os.FileMode, error) {
-	return ioutil.NopCloser(bytes.NewReader(blob.Content)), int64(len(blob.Content)), blob.Mode.Perm(), nil
+	if err := checkFileType(blob.Mode); err != nil {
+		return nil, 0, os.FileMode(0), err
+	}
+	return ioutil.NopCloser(bytes.NewReader(blob.Content)), int64(len(blob.Content)), blob.Mode, nil
+}
+
+func checkFileType(mode os.FileMode) error {
```
Comment (MiguelPires) [inline osutil/syncdir.go:99]: I don't think it makes sense to put this into a function. Without the context of the struct, it's not obvious why this is an error in itself. It also doesn't make the code more concise at the caller level

**Review by MiguelPires — COMMENTED** [review]
Sorry for the delay, one more comment

File: osutil/syncdir.go (hunk 10 above)
Comment (ZeyadYasser) [reply osutil/syncdir.go:99]: updated, thanks for the review!

**Review by MiguelPires — APPROVED** [review]
lgtm, thanks
